File: src/CurveAlign_CT-FIRE/FindConnectedPts.m
function [ con_pts ] = FindConnectedPts(boundaryMask, idx, num )
% FindConnectedPts.m - Find list of connected points around pt, list will be num long
%   if num forces us to look outside the image, then the function
%       returns with all zeros in the list
%   if the first point is not on the mask, then return with all zeros
%
% Inputs
%   boundaryMask = list [row,col] of foreground pixels in the image
%   idx is the index to a point on one of the outlines [row col]
%   num = number of points to return (should be odd)
%
% Optional Inputs
%
% Outputs
%   con_pts = list of connected points (row,col)
% 
% Notes
%   boundaryMask is a 2D image of outlines of a binary mask created in FIJI
%   there are never diagonally connected points in the outline
%   each on pixel should have exactly 2 neighbors
%
% By Jeremy Bredfeldt Laboratory for Optical and
% Computational Instrumentation 2013

con_pts = nan(num,2);

%place pt in the middle of output list (odd length)
hnum = (num-1)/2;
mid = hnum + 1;
con_pts(mid,:) = boundaryMask(idx,:);
%now fill list to begining
sidx = idx; %starting index
visitedList = zeros(1,length(boundaryMask));%keep track of which pixels we've checked
for (i = hnum:-1:1)
    visitedList(idx) = 1;
    idx = GetFirstNeighbor(boundaryMask,idx,visitedList,1);
    con_pts(i,:) = boundaryMask(idx,:);
end

%now fill list in other direction
idx = sidx;
prevIdx = 0;
for (i = mid+1:num)
    visitedList(idx) = 1;
    idx = GetFirstNeighbor(boundaryMask,idx,visitedList,2);
    con_pts(i,:) = boundaryMask(idx,:);
end


end

File: pycurvelets/relativeAngle/find_connected_pts.py
import numpy as np
from get_first_neighbor import get_first_neighbor


def find_connected_pts(boundary_coords, idx, num):
    """
    Find list of `num` connected points around the given index in `boundary_coords`.
    Equivalent to MATLAB's FindConnectedPts.

    Parameters:
        boundary_coords: np.ndarray of shape (N, 2) — (row, col) pairs of outline
        idx: int — index into boundary_coords
        num: int — number of connected points to return (must be odd)

    Returns:
        con_pts: np.ndarray of shape (num, 2) — connected points around center
    """
    con_pts = np.full((num, 2), np.nan)
    hnum = (num - 1) // 2
    mid = hnum
    con_pts[mid] = boundary_coords[idx]

    visited = np.zeros(len(boundary_coords), dtype=bool)
    sidx = idx

    # Fill before center
    for i in range(mid - 1, -1, -1):
        visited[idx] = True
        idx = get_first_neighbor(boundary_coords, idx, visited, direction=1)
        if idx is None:
            return np.full((num, 2), np.nan)
        con_pts[i] = boundary_coords[idx]

    # Fill after center
    idx = sidx
    for i in range(mid + 1, num):
        visited[idx] = True
        idx = get_first_neighbor(boundary_coords, idx, visited, direction=2)
        if idx is None:
            return np.full((num, 2), np.nan)
        con_pts[i] = boundary_coords[idx]

    return con_pts
